writer/writer.go

```diff
@@ -15,27 +15,380 @@
+	if err != nil {
+		return err
+	}
+	defer artifact.Close()
+
+	// we need to ensure correct ordering of files
+	var artifactContent []ReadArchiver
+
+	aInfo := NewJSONStreamArchiver(av.getInfo(), "info")
+	artifactContent = append(artifactContent, aInfo)
+	aHdr := NewFileArchiver(filepath.Join(av.updateLocation, "header.tar.gz"), "header.tar.gz")
+	artifactContent = append(artifactContent, aHdr)
+
+	for _, artifact := range files {
+		bucket, ok := av.updates[artifact.Name()]
+
+		if !ok {
+			return errors.New("artifacts writer: can not find data file")
+		}
+		aData := NewFileArchiver(bucket.archivedPath, "data/0000.tar.gz")
+		artifactContent = append(artifactContent, aData)
+	}
+	return av.writeArchive(artifact, artifactContent, false)
+}
+
+func (av *ArtifactsWriter) storeFile(bucket *updateBucket, upd updateArtifact) error {
+	bucket.files.Files =
+		append(bucket.files.Files, metadata.MetadataFile{File: upd.name})
+	return nil
+}
+
+func (av *ArtifactsWriter) processUpdateBucket(bucket string) error {
+	// get list of update files
+	// at this point we know that `data` exists and contains update(s)
+	bucketLocation := filepath.Join(av.updateLocation, bucket)
+	dataLocation := filepath.Join(av.updateLocation, bucket, "data")
+
+	updateFiles, err := ioutil.ReadDir(dataLocation)
+	if err != nil {
+		return err
+	}
+
+	updBucket := updateBucket{}
+	// iterate through all data files
+	for _, file := range updateFiles {
+		upd := updateArtifact{
+			name:         file.Name(),
+			path:         filepath.Join(dataLocation, file.Name()),
+			updateBucket: bucket,
+			info:         file,
+		}
+		// generate checksums
+		err = av.calculateChecksum(&upd)
+		if err != nil {
+			return err
+		}
+		// TODO: generate signatures
+
+		// store `file` data
+		if err = av.storeFile(&updBucket, upd); err != nil {
+			return err
+		}
+		updBucket.updateArtifacts = append(updBucket.updateArtifacts, upd)
+		updBucket.location = bucket
+		updBucket.path = bucketLocation
+	}
+
+	// move (and compress) updates from `data` to `../data/location.tar.gz`
+	if err = av.archiveData(&updBucket); err != nil {
+		return err
+	}
+
+	av.updates[bucket] = updBucket
+	return nil
+}
+
+// Write writes Mender artifacts metadata compressed archive
+func (av *ArtifactsWriter) Write() error {
+	// get directories list containing updates
+	entries, err := ioutil.ReadDir(av.updateLocation)
+	if err != nil {
+		return err
+	}
+	// iterate through all directories containing updates
+	for _, location := range entries {
+		// check files and directories consistency
+		err = av.headerStructure.CheckHeaderStructure(
+			filepath.Join(av.updateLocation, location.Name()))
+		if err != nil {
+			return err
+		}
+		if err = av.processUpdateBucket(location.Name()); err != nil {
+			return err
+		}
+	}
+	// create compressed header; the intermediate step is needed as we
+	// can not create tar archive containing files compressed on the fly
```

Comment:
They are, and in many cases we are doing thing on the fly. See `writeArchive()`. Here we are writing a file/stream to gzip (using `io.Copy()`) which is then written to tar which is written to file. 
Here the limitation is that we need to write tar header before writing content. And part of header is size of file which will be written. So we need to have a file in advance to check size. 
Maybe there is some work-around possible, but since this is ONLY for writing compressed headers (pretty small) and will be done in Yocto only I don't want to spend time fixing it.
FYI reader won't store any intermediate files. It will read everything on the fly.